File: android/src/main/kotlin/africa/ejara/trustdart/coins/BSC.kt
import africa.ejara.trustdart.Coin
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toHexByteArray
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Ethereum.SigningOutput
import wallet.core.jni.proto.Ethereum


class BSC : Coin("BSC", CoinType.SMARTCHAIN) {

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val signingInput = Ethereum.SigningInput.newBuilder()
        val wallet = HDWallet(mnemonic, passphrase)
        val private_key = wallet.getKey(coinType, path)
        signingInput.apply {
            privateKey = ByteString.copyFrom(private_key.data())
            toAddress = txData["toAddress"] as String
            chainId = ByteString.copyFrom((txData["chainID"] as String).toHexByteArray())
            nonce = ByteString.copyFrom((txData["nonce"] as String).toHexByteArray())
            gasPrice = ByteString.copyFrom((txData["gasPrice"] as String).toHexByteArray())
            gasLimit = ByteString.copyFrom((txData["gasLimit"] as String).toHexByteArray())
            transaction = Ethereum.Transaction.newBuilder().apply {
                transfer = Ethereum.Transaction.Transfer.newBuilder().apply {
                    amount = ByteString.copyFrom((txData["amount"] as String).toHexByteArray())
                }.build()
            }.build()
        }

        val sign = AnySigner.sign(signingInput.build(), coinType, SigningOutput.parser())
        return "0x" + Numeric.toHexString(sign.encoded.toByteArray())

    }
}

File: ios/Classes/coins/BSC.swift
import WalletCore

class BSC: Coin  {
    init() {
        super.init(name: "BSC", coinType: .smartChain)
    }

    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)

        let input = EthereumSigningInput.with {
            $0.chainID = Data(hexString: (txData["chainID"] as! String))!
            $0.nonce = Data(hexString: (txData["nonce"] as! String))! 
            $0.gasPrice = Data(hexString: (txData["gasPrice"] as! String))! 
            $0.gasLimit = Data(hexString: (txData["gasLimit"] as! String))!
            $0.toAddress = txData["toAddress"] as! String
            $0.privateKey = privateKey!.data
            $0.transaction = EthereumTransaction.with {
                $0.transfer = EthereumTransaction.Transfer.with {
                    $0.amount = Data(hexString: (txData["amount"] as! String))!
                }
            }
        }

        let sign: EthereumSigningOutput = AnySigner.sign(input: input, coin: coinType)
        return "0x"+sign.encoded.hexString

    }
}
